Use default system shell to capture shell output.

pavoljuhas opened the pull request:
Adhere to the user `$SHELL` preference when capturing output from system
commands.  This makes the getoutput() and system() interactive functions
behave in a consistent way.  It also makes user shell customizations, i.e.,
aliases and non-exported shell variables, available within the captured 
command.

Example of an inconsistent behavior when `SHELL=/bin/zsh`:

```
In [1]: !print -l a b
a
b

In [2]: out = !print -l a b

In [3]: out
Out[3]: ['/bin/sh: print: command not found']
```

pavoljuhas committed:
Use the default shell to capture shell output.

Adhere to the user $SHELL choice when capturing output from system
commands.  This makes the getoutput() behavior consistent with
that of the system() interactive function.

takluyver commented on IPython/utils/_process_common.py:
```diff
@@ -70,6 +71,7 @@ def process_handler(cmd, callback, stderr=subprocess.PIPE):
     # On win32, close_fds can't be true when using pipes for stdin/out/err
     close_fds = sys.platform != 'win32'
     p = subprocess.Popen(cmd, shell=isinstance(cmd, py3compat.string_types),
+                         executable=os.environ.get('SHELL'),
```
This should only be set if we're using the shell - which, from the line above, is the case if `cmd` is a string (...I don't like this code).

I think we should also only be using this on Unix systems.

pavoljuhas committed:
Run user $SHELL only for shell commands.

Do this only for POSIX systems.  Use default shell execution on
Windows even if SHELL is set in the environment.

pavoljuhas commented:
@takluyver - good points, please see the updated PR.

takluyver commented:
Thanks, looks OK to me.
